crates/apollo-compiler/src/validation/subscription.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{
+    diagnostics::IntrospectionField, hir::Selection, ApolloDiagnostic, FileId, ValidationDatabase,
+};
+
+pub fn check(db: &dyn ValidationDatabase, file_id: FileId) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+    let subscription_operations = db.upcast().subscription_operations(file_id);
+
+    // Subscription fields must not have an introspection field in the selection set.
+    //
+    // Return an Introspection Field error in case of an introspection field existing as the root field in the set.
+    for op in subscription_operations.iter() {
+        for selection in op.selection_set().selection() {
+            if let Selection::Field(field) = selection {
+                if field.name().starts_with("__") {
```

Comment:
it's only a convention to keep `__` for introspection fields only, so theoretically someone can just have a field that starts with `__` in their schema. It'd be best to check for `__typename`, `__schema` and `__type` directly. 

If you rebase from main, we've just merged #421, so you should be able to use `field.is_introspection()` here.

Reply:
Ah, good to know! I'll update this and use that great `field.is_introspection()` method 👍